soroban-sdk-tools-macro/src/error.rs

```diff
@@ -1,30 +1,678 @@
+
+fn expand_scerr_root(input: &DeriveInput, infos: &[VariantInfo]) -> syn::Result<proc_macro2::TokenStream> {
+    let name = &input.ident;
+    let has_data = infos.iter().any(|i| i.field_ty.is_some());
+    
+    let enum_def = generate_enum_def(name, infos, has_data);
+    let contract_error_impl = generate_contract_error_impl_root(name, infos);
+    let from_impls = generate_from_trait_impls(name, infos)?;
+    let invoke_impl = generate_invoke_error_impl(name, infos);
+    let soroban_impls = generate_soroban_conversions(name);
+
+    // Support for ?? on Result<T, InvokeError> where T is an inner error (FCC)
+    let double_q_impls: Vec<_> = infos.iter().filter(|i| i.is_fcc).map(|info| {
+        let ident = info.ident;
+        let ty = info.field_ty.as_ref().unwrap();
+        let ok_construct = if let Some(field) = &info.field_ident {
+            quote! { Self::#ident { #field: inner } }
+        } else {
+            quote! { Self::#ident(inner) }
+        };
+        quote! {
+            impl From<Result<#ty, soroban_sdk::InvokeError>> for #name {
+                fn from(res: Result<#ty, soroban_sdk::InvokeError>) -> Self {
+                    match res {
+                        Ok(inner) => #ok_construct,
+                        Err(e) => Self::from(e),
+                    }
+                }
+            }
+        }
+    }).collect();
+
+    Ok(quote! {
+        #enum_def
+        #contract_error_impl
+        #from_impls
+        #invoke_impl
+        #soroban_impls
+        #(#double_q_impls)*
+    })
+}
+
+fn expand_scerr_basic(input: &DeriveInput, infos: &[VariantInfo]) -> syn::Result<proc_macro2::TokenStream> {
+    let name = &input.ident;
+    let variants: Vec<_> = infos.iter().map(|i| {
+        let id = i.ident;
+        let c = i.code;
+        quote! { #id = #c }
+    }).collect();
+    
+    let from_arms: Vec<_> = infos.iter().map(|i| {
+        let c = i.code;
+        let ident = i.ident;
+        quote! { #c => Some(#name::#ident) }
+    }).collect();
+
+    let desc_arms: Vec<_> = infos.iter().map(|i| {
+        let ident = i.ident;
+        let d = &i.description;
+        quote! { #name::#ident => #d }
+    }).collect();
+
+    Ok(quote! {
+        #[soroban_sdk::contracterror]
+        #[repr(u32)]
+        #[derive(Copy, Clone, Debug, Eq, PartialEq)]
+        pub enum #name {
+            #(#variants),*
+        }
+
+        impl soroban_sdk_tools::error::ContractError for #name {
+            fn into_code(self) -> u32 { self as u32 }
+            fn from_code(code: u32) -> Option<Self> {
+                match code {
+                    #(#from_arms,)*
+                    _ => None,
+                }
+            }
+            fn description(&self) -> &'static str {
+                match *self {
+                    #(#desc_arms,)*
+                }
+            }
+        }
+    })
+}
+
+pub fn scerr_impl(attr: TokenStream, item: TokenStream) -> TokenStream {
```

Comment:
I is less noisy if there is another intermediate method with `Result<proc_macro2::TokenStream, syn::Error>`

Then this method is just parsing then calling the other.

```rust
let mode = match parse_scerr_attr(attr) {
    Ok(m) => m,
    Err(e) => return e.to_compile_error().into(),
};
let input = parse_macro_input!(item as DeriveInput); // This one panics I believe.
match get_expanded(mode, input) {
        Ok(ts) => ts.into(),
        Err(e) => e.to_compile_error().into(),
    }
```

Then in `get_expanded` or whatever you name it, you can use normal `?`.

And then that method can be tested with `syn` types.